carbon-design-system/carbon-addons-iot-react
# fix(tablereducer): SelectAll true if all rows are selected on initial render

Closes #3651 

**Summary**

- Select all checkbox has state `true` if all rows are selected

**Change List (commits, features, bugs, etc)**

- Change `isSelectAllSelected` to `true` if `selectedIds` has the same length as all rows (including nesting)
- Change `tableRegister` function signature in order to check `hasRowSelection === 'multi'`
- Add new test
- Update test

**Acceptance Test (how to verify the PR)**

- Go to story [Table / Playground](https://deploy-preview-3652--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-table--playground)
- In the `Nesting & expansion` knobs set `hasRowExpansion: true` and `hasRowNesting: true`
- In the `Selections & actions` knobs:
  - Select `useRadioButtonsingleSelect`
  - Change value `hasRowSelection: single` and then back to `hasRowSelection: multi` to re-render the table
  - Paste code snippet below to `view.table.selectedIds`.
- Verify that Select all checkbox is selected

```js
[
    "row-0",
    "row-1",
    "row-1_A",
    "row-1_B",
    "row-2",
    "row-2_A",
    "row-2_B",
    "row-3",
    "row-3_A",
    "row-3_B",
    "row-4",
    "row-4_A",
    "row-4_B",
    "row-4_B-1",
    "row-4_B-2",
    "row-4_B-2-A",
    "row-4_B-2-B",
    "row-4_B-3",
    "row-4_C",
    "row-4_D",
    "row-4_D-1",
    "row-4_D-2",
    "row-4_D-3",
    "row-5",
    "row-5_A",
    "row-5_B",
    "row-6",
    "row-6_A",
    "row-6_B",
    "row-7",
    "row-7_A",
    "row-7_B",
    "row-8",
    "row-9",
    "row-9_A",
    "row-9_B",
    "row-10",
    "row-10_A",
    "row-10_B",
    "row-11",
    "row-11_A",
    "row-11_B",
    "row-12",
    "row-13",
    "row-13_A",
    "row-13_B",
    "row-14",
    "row-14_A",
    "row-14_B",
    "row-15",
    "row-15_A",
    "row-15_B",
    "row-16",
    "row-17",
    "row-17_A",
    "row-17_B",
    "row-18",
    "row-18_A",
    "row-18_B",
    "row-19",
    "row-19_A",
    "row-19_B",
    "row-20",
    "row-21",
    "row-21_A",
    "row-21_B",
    "row-22",
    "row-22_A",
    "row-22_B",
    "row-23",
    "row-23_A",
    "row-23_B",
    "row-24",
    "row-25",
    "row-25_A",
    "row-25_B",
    "row-26",
    "row-26_A",
    "row-26_B",
    "row-27",
    "row-27_A",
    "row-27_B",
    "row-28",
    "row-29",
    "row-29_A",
    "row-29_B",
    "row-30",
    "row-30_A",
    "row-30_B",
    "row-31",
    "row-31_A",
    "row-31_B",
    "row-32",
    "row-33",
    "row-33_A",
    "row-33_B",
    "row-34",
    "row-34_A",
    "row-34_B",
    "row-35",
    "row-35_A",
    "row-35_B",
    "row-36",
    "row-37",
    "row-37_A",
    "row-37_B",
    "row-38",
    "row-38_A",
    "row-38_B",
    "row-39",
    "row-39_A",
    "row-39_B",
    "row-40",
    "row-41",
    "row-41_A",
    "row-41_B",
    "row-42",
    "row-42_A",
    "row-42_B",
    "row-43",
    "row-43_A",
    "row-43_B",
    "row-44",
    "row-45",
    "row-45_A",
    "row-45_B",
    "row-46",
    "row-46_A",
    "row-46_B",
    "row-47",
    "row-47_A",
    "row-47_B",
    "row-48",
    "row-49",
    "row-49_A",
    "row-49_B",
    "row-50",
    "row-50_A",
    "row-50_B",
    "row-51",
    "row-51_A",
    "row-51_B",
    "row-52",
    "row-53",
    "row-53_A",
    "row-53_B",
    "row-54",
    "row-54_A",
    "row-54_B",
    "row-55",
    "row-55_A",
    "row-55_B",
    "row-56",
    "row-57",
    "row-57_A",
    "row-57_B",
    "row-58",
    "row-58_A",
    "row-58_B",
    "row-59",
    "row-59_A",
    "row-59_B",
    "row-60",
    "row-61",
    "row-61_A",
    "row-61_B",
    "row-62",
    "row-62_A",
    "row-62_B",
    "row-63",
    "row-63_A",
    "row-63_B",
    "row-64",
    "row-65",
    "row-65_A",
    "row-65_B",
    "row-66",
    "row-66_A",
    "row-66_B",
    "row-67",
    "row-67_A",
    "row-67_B",
    "row-68",
    "row-69",
    "row-69_A",
    "row-69_B",
    "row-70",
    "row-70_A",
    "row-70_B",
    "row-71",
    "row-71_A",
    "row-71_B",
    "row-72",
    "row-73",
    "row-73_A",
    "row-73_B",
    "row-74",
    "row-74_A",
    "row-74_B",
    "row-75",
    "row-75_A",
    "row-75_B",
    "row-76",
    "row-77",
    "row-77_A",
    "row-77_B",
    "row-78",
    "row-78_A",
    "row-78_B",
    "row-79",
    "row-79_A",
    "row-79_B",
    "row-80",
    "row-81",
    "row-81_A",
    "row-81_B",
    "row-82",
    "row-82_A",
    "row-82_B",
    "row-83",
    "row-83_A",
    "row-83_B",
    "row-84",
    "row-85",
    "row-85_A",
    "row-85_B",
    "row-86",
    "row-86_A",
    "row-86_B",
    "row-87",
    "row-87_A",
    "row-87_B",
    "row-88",
    "row-89",
    "row-89_A",
    "row-89_B",
    "row-90",
    "row-90_A",
    "row-90_B",
    "row-91",
    "row-91_A",
    "row-91_B",
    "row-92",
    "row-93",
    "row-93_A",
    "row-93_B",
    "row-94",
    "row-94_A",
    "row-94_B",
    "row-95",
    "row-95_A",
    "row-95_B",
    "row-96",
    "row-97",
    "row-97_A",
    "row-97_B",
    "row-98",
    "row-98_A",
    "row-98_B",
    "row-99",
    "row-99_A",
    "row-99_B"
]
```

**Regression Test (how to make sure this PR doesn't break old functionality)**

- tests here



**Things to look for during review**
